File: design/1357_apply_discount_every_n_orders/1357_apply_discount_every_n_orders.py

class Solution(object):
    def __init(self, n, discount, products, prices):
        self.price = {p : prices[i] for i,p in enumerate(products)}
        self.discount = float(discount)
        self.n = n
        self.cnt = 0

    def getBill(self, product, amount):
        self.cnt += 1
        total = 0
        for i,p in enumerate(product):
            total += self.price[p] * amount[i]
        
        if self.cnt % self.n == 0 :
            total = total * (1 - self.discount/100)
        return total

File: design/1357_apply_discount_every_n_orders/1357_apply_discount_every_n_orders.go
package main

type Cashier struct {
	count         int
	discount      int
	discountCount int
	prices        map[int]int
}

func Constructor(n int, discount int, products []int, prices []int) Cashier {
	c := Cashier{
		count:         0,
		discount:      discount,
		discountCount: n,
		prices:        map[int]int{},
	}

	for i := 0; i < len(products); i++ {
		c.prices[products[i]] = prices[i]
	}
	return c
}

func (this *Cashier) GetBill(product []int, amount []int) (rtn float64) {
	this.count++
	total := 0
	for i := 0; i < len(product); i++ {
		total += this.prices[product[i]] * amount[i]
	}
	if this.count%this.discountCount == 0 {
		rtn = float64(total) * float64(100-this.discount) / 100
	} else {
		rtn = float64(total)
	}
	return
}
